pkg/operator/ceph/cluster/controller.go

```diff
@@ -261,58 +266,117 @@ func (r *ReconcileCephCluster) reconcile(request reconcile.Request) (reconcile.R
 
 	// DELETE: the CR was deleted
 	if !cephCluster.GetDeletionTimestamp().IsZero() {
-		doCleanup := true
-		logger.Infof("deleting ceph cluster %q", cephCluster.Name)
+		return r.reconcileDelete(cephCluster)
+	}
 
-		// Start cluster clean up only if cleanupPolicy is applied to the ceph cluster
-		stopCleanupCh := make(chan struct{})
-		if cephCluster.Spec.CleanupPolicy.HasDataDirCleanPolicy() && !cephCluster.Spec.External.Enable {
-			// Set the deleting status
-			updateStatus(r.client, request.NamespacedName, cephv1.ConditionDeleting)
+	// Do reconcile here!
+	ownerInfo := k8sutil.NewOwnerInfo(cephCluster, r.scheme)
+	if err := r.clusterController.onAdd(cephCluster, ownerInfo); err != nil {
+		return reconcile.Result{}, cephCluster, errors.Wrapf(err, "failed to reconcile cluster %q", cephCluster.Name)
+	}
 
-			monSecret, clusterFSID, err := r.clusterController.getCleanUpDetails(cephCluster.Namespace)
-			if err != nil {
-				logger.Warningf("failed to get mon secret. Skip cluster cleanup and remove finalizer. %v", err)
-				doCleanup = false
-			}
+	// Return and do not requeue
+	return reconcile.Result{}, cephCluster, nil
+}
 
-			if doCleanup {
-				cephHosts, err := r.clusterController.getCephHosts(cephCluster.Namespace)
-				if err != nil {
-					return reconcile.Result{}, cephCluster, errors.Wrapf(err, "failed to find valid ceph hosts in the cluster %q", cephCluster.Namespace)
-				}
-				go r.clusterController.startClusterCleanUp(stopCleanupCh, cephCluster, cephHosts, monSecret, clusterFSID)
-			}
+// refresh the ceph cluster with the latest changes in-place
+func (r *ReconcileCephCluster) refreshCephCluster(c **cephv1.CephCluster) error {
```

Comment:
How about returning the new one instead of a double pointer? Double pointers are an evil pattern IMO. :)
```suggestion
func (r *ReconcileCephCluster) refreshCephCluster(c *cephv1.CephCluster) (*cephv1.CephCluster, error) {
```

Reply:
I will fix this. For clarity, I think it will be just as easy to not have the helper at that point, because what this helps with (with the double pointer) is to not overwrite the old `cephCluster` var if the refresh fails. I was trying to save some more lines of copy-pasting and making the error checking easier in the call to this function.